tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+package at.forsyte.apalache.io.itf
+
+import at.forsyte.apalache.io.itf.ItfToTla._
+import at.forsyte.apalache.io.json.{JsonDeserializationError, JsonRepresentation, ScalaFactory}
+import at.forsyte.apalache.io.lir.Trace
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.types.parser.DefaultType1Parser
+import at.forsyte.apalache.tla.types.{tla, ModelValueHandler}
+
+/**
+ * Reads a sequence of States, i.e. mappings from variable names to TlaEx values, from an ITF json.
+ *
+ * @tparam T
+ *   Any class extending JsonRepresentation
+ * @author
+ *   Jure Kukovec
+ */
+class ItfToTla[T <: JsonRepresentation](
+    scalaFactory: ScalaFactory[T]) {
+
+  /**
+   * Checks that the Apalache-specific ITF requirements are met, as defined in ADR15. If yes, returns a mapping from
+   * variables to their types.
+   */
+  private[itf] def validateShapeAndGetTypes(json: T): Map[String, TlaType1] = {
+    val varTypesOpt = for {
+      meta <- json.getFieldOpt(META_FIELD)
+      varTypes <- meta.getFieldOpt(VAR_TYPES_FIELD)
+    } yield varTypes
+    val varTypesJson = varTypesOpt.getOrElse {
+      throw new JsonDeserializationError("ITF JSON does not satisfy Apalache requirements. " +
+        "Toplevel \"#meta\" field must exist and contain \"varTypes\". " +
+        "See ADR15: https://apalache.informal.systems/docs/adr/015adr-trace.html#itf-as-input-to-apalache. ")
+    }
```

Comment:
> I left just a few initial comments, but I wanted to propose this alternative approach:
> 
> I believe that if we define a representation of the ITF format in Scala classes, we can serialization and deserialization more or less for free via ujson and upickle.
> 
> This could give us an approach that covered this new use case but resulted in overall reduction of code.
> 
> We'd end up with two mapping `TlaModule -> ITF` and `ITF -> TlaModule`, and then the de/serialization of the `ITF` type itself should be automated.
> 
> Any thoughts on this alternative?

Two, mainly. Firstly, we already have an in-Scala representation: `IndexedSeq[Map[String, TlaEx]]` (`lir.Trace`). Is is the only real programmatic representation of a trace you need IMO.
Secondly, the problem I have with outsourcing (de)serialization, is that you use any sort of useful error message feedback; Assume you have a more complicated Scala `ITF` class and you want to use a library to automatically deserialize it from JSON. Because you're outsourcing, the only error message you can get is `SomeException("Input file malformed.")`. It won't tell you, for example, that your JSON declares 3 values in `vars`, but the states only have 2 variables, because that's not a JSON formatting issue, that's a logical constraint not being satisfied.